Pullrequest for Nbaars

commjoen opened the pull request:
Nbaars can use this PR to comment on code found

jordy1111100 committed:
Merge pull request #60 from xebia-research/feature/Implement_seed_method_#24

Feature/implement seed method #24

nikro412956 committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/Analyse_headers_#13

# Conflicts:
#	src/main/java/com/xebia/headerbuddy/models/entities/repositories/EheaderRepository.java

Commit:
Made a begin with checking the db for a api key

OldaranT committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/Analyse_headers_#13

# Conflicts:
#	src/main/java/com/xebia/headerbuddy/models/entities/repositories/EheaderRepository.java

anthony0tran committed:
Implemented shell script.

Call API from console.

jordy1111100 committed:
Clean branch and merged with develop

jordy1111100 committed:
Fixed bug by deleting Analyser class

OldaranT committed:
Created static methode to convert Header to Eheader.

anthony0tran committed:
Merge branch 'develop' into feature/Implement_CLI_#31

anthony0tran committed:
Merge pull request #61 from xebia-research/feature/Implement_CLI_#31

Feature/implement cli #31

nikro412956 committed:
set up the basic anylizer.
added the function that organizes the headers from the database in 3 catagories.
added the function that organizes the found headers in 3 catagoeries.
added the funtions for returning the orginezed Sets

nikro412956 committed:
Merge branch 'feature/Analyse_headers_#13' of https://github.com/xebia-research/headerbuddy into feature/Analyse_headers_#13

# Conflicts:
#	src/main/java/com/xebia/headerbuddy/models/Analyzer.java
#	src/main/java/com/xebia/headerbuddy/models/entities/repositories/EheaderRepository.java

OldaranT committed:
Merge branch 'feature/Analyse_headers_#13' of https://github.com/xebia-research/headerbuddy into feature/Analyse_headers_#13

# Conflicts:
#	src/main/java/com/xebia/headerbuddy/controllers/HeaderBuddyController.java

jordy1111100 committed:
Bug fix, added bean end2end test

jordy1111100 committed:
Merge branch 'feature/Analyse_headers_#13' of https://github.com/xebia-research/headerbuddy into feature/Analyse_headers_#13

# Conflicts:
#	checkerReport/dependency-check-report.html

anthony0tran committed:
added the -i (IP) and -p (PORT) flag to the script

jordy1111100 committed:
Fixed wrong properties settings

anthony0tran committed:
minor typo fix

anthony0tran committed:
Merge branch 'develop' into feature/Implement_CLI_#31

Commit:
API keys are added to db with proper errors

Commit:
api key in request now is validated

jordy1111100 committed:
Now detecting missing do headers and don't headers

jordy1111100 committed:
Now able to create list with found do headers

MichelV96 committed:
Spider class added

jordy1111100 committed:
Now able to detect recommended headers

anthony0tran committed:
Merge pull request #62 from xebia-research/feature/Implement_CLI_#31

Feature/implement cli #31

OldaranT committed:
Reformatted the seed methoded with more detail.
Finished Referrer-Policy header.

Commit:
renamed Spider -> WebCrawler, added unit test, added temporary get request for crawl

Commit:
renamed Spider -> WebCrawler, added unit test, added temporary get request for crawl

jordy1111100 committed:
Analyser works

OldaranT committed:
Added new headers with value's in seed

Commit:
fixed unit test

nikro412956 committed:
added some csp header values to the seed method

Commit:
Fixed some minor config, wouldnt get through travis

MichelV96 committed:
Rewrote end 2 end test; can be used with repositories now

jordy1111100 committed:
Modified controller so that report class is no longer necessary and date is now displayed right

MichelV96 committed:
Api key end 2 end test

MichelV96 committed:
sql link

MichelV96 committed:
Travis changes

nikro412956 committed:
added more csp directives

nikro412956 committed:
Merge branch 'feature/Implement_seed_method_#24' of https://github.com/xebia-research/headerbuddy into feature/Implement_seed_method_#24

MichelV96 committed:
pipeline test for e2e and it testing

jordy1111100 committed:
Bugfix duplicated missing do headers

Commit:
moved unit test to integration test, added some integration tests as well

nikro412956 committed:
did a reformat on the project and changed the ruleset for checkstyle to remove false positives

nikro412956 committed:
finished the csp header directives part of the seed

MichelV96 committed:
Pipeline test for e2e and it test

OldaranT committed:
Created analzyer structure for single responsibilty and for more flexibel code.
Fixed some PMD Notifications.

Commit:
integrated crawler with main call

jordy1111100 committed:
Added X-Power-By header and default user

nikro412956 committed:
fine tuned the rulesets for checkstyle and pmd.
also reënabled the fail on violation.

MichelV96 committed:
Merge branch 'develop' into feature/api_key_implementation_#25

MichelV96 committed:
Merge branch 'develop' into feature/api_key_implementation_#25

MichelV96 committed:
Merge branch 'feature/api_key_implementation_#25' of https://github.com/xebia-research/headerbuddy into feature/api_key_implementation_#25

MichelV96 committed:
imports and review adjustments

MichelV96 committed:
Mysql properties

jordy1111100 committed:
Merge pull request #63 from xebia-research/feature/api_key_implementation_#25

Feature/api key implementation #25

jordy1111100 committed:
Bugfix for doAnalyse and checkstyle adjustments

Commit:
merge with develop

MichelV96 committed:
Intergration test maven exec

jordy1111100 committed:
Merge branch 'develop' into feature/Analyse_headers_#13

# Conflicts:
#	checkerReport/dependency-check-report.html
#	src/main/java/com/xebia/headerbuddy/controllers/HeaderBuddyController.java
#	src/main/java/com/xebia/headerbuddy/models/entities/Euser.java
#	src/main/java/com/xebia/headerbuddy/models/entities/repositories/EuserRepository.java
#	src/test/java/com/xebia/headerbuddy/e2e/HeaderBuddyControllerTest.java
#	src/test/java/com/xebia/headerbuddy/e2e/HeaderBuddyOutputTest.java
#	src/test/java/com/xebia/headerbuddy/e2e/HeaderBuddyUrlParamTest.java

MichelV96 committed:
mysql properties

MichelV96 committed:
Travis debugging

MichelV96 committed:
Travis debugging

anthony0tran committed:
added docker logs to check why the application container crashes

jordy1111100 committed:
Created some unit test for analyseHandlers and removed unused method

jordy1111100 committed:
wrong proporties fix

anthony0tran committed:
Used wrong container name.

anthony0tran committed:
add before script again to prevent mysql from initializing

OldaranT committed:
Added  Last-Modified,  X-XSS-Protection,  X-Content-Type-Options,  Cache-Control, Set-Cookie, Pragma headers in seed.

MichelV96 committed:
Travis debug

MichelV96 committed:
Merge branch 'feature/Web_Crawler_#9' of https://github.com/xebia-research/headerbuddy into feature/Web_Crawler_#9

MichelV96 committed:
Travis debug

MichelV96 committed:
Travis debug

MichelV96 committed:
Maven changes for verify

MichelV96 committed:
Merge pull request #65 from xebia-research/feature/Web_Crawler_#9

Feature/web crawler #9

jordy1111100 committed:
Added tests for analyzers

jordy1111100 committed:
Merged with develop

anthony0tran committed:
added -a request API key flag

nikro412956 committed:
added the the required components so travis wil push to google cloud storage

nikro412956 committed:
this should fix the problem where the branch is not permitted to deploy.
i removed the "on" tag

nikro412956 committed:
redid the deploy section to see if it will deploy this time

nikro412956 committed:
using environmental variables instead of branch name to not constantly change the branch name in the .yml file.
added skip cleanup to the deploy to keep the artifacts

nikro412956 committed:
removed $pull_request variable from the file

jordy1111100 committed:
Changed analyse structure

jordy1111100 committed:
Changed Analyse test

MichelV96 committed:
Merge pull request #66 from xebia-research/feature/Analyse_headers_#13

Feature/analyse headers #13

OldaranT committed:
Added new headers in the seed methode: Set-Cookie, Pragma and Vary.

jordy1111100 committed:
merged implement seed method with develop

OldaranT committed:
Merge branch 'feature/Implement_seed_method_#24' of https://github.com/xebia-research/headerbuddy into feature/Implement_seed_method_#24

anthony0tran committed:
Merge pull request #67 from xebia-research/feature/Implement_seed_method_#24

Feature/implement seed method #24

anthony0tran committed:
added api key to shell script

anthony0tran committed:
Merge branch 'develop' into feature/Implement_CLI_#31

MichelV96 committed:
Checkstyle and pmd fixes + enabled

OldaranT committed:
Added  the following headers to the seed method: Public-Key-Pins, Server-Timing and X-Runtime

MichelV96 committed:
Merge branch 'develop' into feature/adding_the_xebia_rulesets_#54

jordy1111100 committed:
Url's now integrated with report

nikro412956 committed:
added basic logging support

jordy1111100 committed:
Merge pull request #68 from xebia-research/feature/Implement_CLI_#31

Feature/implement cli #31

OldaranT committed:
Merge branch 'develop' into feature/adding_the_xebia_rulesets_#54

OldaranT committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/Implement_seed_method_#24

MichelV96 committed:
Merge pull request #69 from xebia-research/feature/adding_the_xebia_rulesets_#54

Feature/adding the xebia rulesets #54

jordy1111100 committed:
User in report can not be null anymore

OldaranT committed:
Merge branch 'develop' into feature/Implement_seed_method_#24

OldaranT committed:
Corrected some spelling.

OldaranT committed:
Merge pull request #70 from xebia-research/feature/Implement_seed_method_#24

Feature/implement seed method #24

nikro412956 committed:
moved the logging setting to its own xml file

OldaranT committed:
Merge branch 'release/2.0.0'

OldaranT committed:
Merge tag '2.0.0' into develop

jordy1111100 committed:
advice is now sorted

nikro412956 committed:
set the datasource back to the right url

jordy1111100 committed:
Resolved merge conflicts

jordy1111100 committed:
Checkstyle cleanup

MichelV96 committed:
Merge pull request #71 from xebia-research/feature/Create_report_#17

Feature/create report #17

OldaranT committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/Detect_type_of_service_#10

MichelV96 committed:
Swagger intergration and configuration

nikro412956 committed:
added the beginning of the logging messages.
TODO:
make a logging file instead of logging to the console

MichelV96 committed:
Merge with develop

nikro412956 committed:
changed the datasource back to the docker ccompose version

MichelV96 committed:
Swagger Config file

OldaranT committed:
Created Protocolhandler to define what protocols are found in the requested url.

OldaranT committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into develop

nikro412956 committed:
the log will now be displayed in the console as well as in a file.

jordy1111100 committed:
Merge pull request #72 from xebia-research/feature/SwaggerIO_#32

Feature/swagger io #32

nikro412956 committed:
changed the location of the log file to the target map.
added an info log to the Report.java class for when you create a Report.

OldaranT committed:
Resolved Checkstyle & findbugs errors.

jordy1111100 committed:
Resolved value description too long issue

jordy1111100 committed:
Merge branch 'develop' into feature/Implement_seed_method_#24

nikro412956 committed:
Support logging #28
renamed the directory for the reports to "globalReports".
added custom file appenders to split the log file into: error-warn-log.log, info-log.log, debug-log.log, trace-log.log.
formatted the log files in such a way that they are human readable.
console output for the log will be level warn and error.
these log files are placed in to a log directory in the globalRepeorts directory.
the log files are not pushed to github.
renamed the checkerRules directory to ruleSets.

jordy1111100 committed:
Removed vulnerable dependacy docker-maven dependency

nikro412956 committed:
Support logging #28
removed the include that pointed towards base.xml because it caused the console to overflow with log statements

jordy1111100 committed:
Fix docker-compose.yml

jordy1111100 committed:
Dockerfile

jordy1111100 committed:
Updated license checker

jordy1111100 committed:
Merge branch 'develop' into feature/Implement_seed_method_#24

# Conflicts:
#	checkerReport/dependency-check-report.html

nikro412956 committed:
Support logging #28
added a logging statement to the creation of a response entity

jordy1111100 committed:
Merge pull request #74 from xebia-research/feature/Implement_seed_method_#24

Feature/implement seed method #24

nikro412956 committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/Support_logging_#28

# Conflicts:
#	globalReports/Third-Party-Licenses
#	globalReports/dependency-check-report.html
#	pom.xml
#	src/main/java/com/xebia/headerbuddy/controllers/HeaderBuddyController.java
#	src/main/java/com/xebia/headerbuddy/models/Report.java

OldaranT committed:
Updated seed method for protcol changes.
Resolved issue where recc where not regonicsed if the value was '*'
Fixed x-csp header.

jordy1111100 committed:
Updated gitignore

OldaranT committed:
Updated git ignore.

OldaranT committed:
Merge branch 'develop' into feature/Detect_type_of_service_#10

# Conflicts:
#	src/main/java/com/xebia/headerbuddy/controllers/HeaderBuddyController.java

nikro412956 committed:
changed the console output to use the default console appender from logback so that it is compatible with docker.
saving the log files to the resource directory

anthony0tran committed:
changed jsoup dependency version to the latest

anthony0tran committed:
added checkerReport

Commit:
Made a start with html report

jordy1111100 committed:
Merge branch 'feature/Detect_type_of_service_#10' of https://github.com/xebia-research/headerbuddy into feature/Detect_type_of_service_#10

MichelV96 committed:
added certificate model; added note to report; added certificate get server certificate;

MichelV96 committed:
Merge branch 'develop' into feature/Certificate_Details_#73

MichelV96 committed:
method getCertificateDetails tests

jordy1111100 committed:
Added profile to report and modified protocolHandler

nikro412956 committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/adding_GCS_from_Xebia_#64

# Conflicts:
#	.travis.yml

OldaranT committed:
Added unit test to Protocalhandler
Includes:
- Http only
- Https only
- Http and Https
- Unkown protocol

nikro412956 committed:
Support logging #28
made custom console appender because otherwise i can not use a filter for it.
renamed the custom-file-appender.xml file to custom-appenders.xml
added error loggers to the catches in the webcrawler
you can now retrieve the log files from the container

nikro412956 committed:
Support logging #28
removed Report.java because it shouldn't be there in the first place

MichelV96 committed:
Merge pull request #77 from xebia-research/feature/Support_logging_#28

Feature/support logging #28

MichelV96 committed:
Merge with develop

Commit:
html output now returns a html report using Rythm engine

Commit:
fixed incorrect properties config

nikro412956 committed:
removed redundant directory's

nikro412956 committed:
redoing the travis build because it failed for an unknown reason

OldaranT committed:
Removed some not used imports.
Cleaned the unit test for protocolhandler.

OldaranT committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/Detect_type_of_service_#10

nikro412956 committed:
added a placeholder so that the log directory is pre generated because it looks like the tests error because there is no log directory yet.

jordy1111100 committed:
MethodHandler now retrieves headers from found url's crawler

jordy1111100 committed:
CHeckstyle cleanup

jordy1111100 committed:
Analyzer is no longer case sensitive

Commit:
added html errors and details to html report

Commit:
cleaned up code with checkstyle

jordy1111100 committed:
made changes based on review

jordy1111100 committed:
Corrected grammar in seed

jordy1111100 committed:
Capitals in seed fix

jordy1111100 committed:
Merge pull request #78 from xebia-research/feature/Detect_type_of_service_#10

Feature/detect type of service #10

anthony0tran committed:
Resolved vulnerabilities in dependencies.

anthony0tran committed:
merged with develop

anthony0tran committed:
Removed commmented code

anthony0tran committed:
Merge pull request #79 from xebia-research/feature/Resolve_vulnerable_dependencies_#75

Feature/resolve vulnerable dependencies #75

nikro412956 committed:
putting the log file directory back in root because otherwise it would generate a variable_not_defined directory.
log files can not be extracted from the docker container and putting the logs into a database does not work at the moment.

MichelV96 committed:
Configurable settings for api key on/off and webcrawler limit

nikro412956 committed:
removed unnecessary property in the custom appenders

Commit:
merge with develop

MichelV96 committed:
Merge remote-tracking branch 'origin/develop' into feature/Support_configurable_settings_#27

nikro412956 committed:
Support logging #28
removed the log files registry because that won't work anymore

nikro412956 committed:
removed the codestyles from the gitignore because it causes an error with git if you do not track it

nikro412956 committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/Support_logging_#28

# Conflicts:
#	globalReports/dependency-check-report.html

nikro412956 committed:
made it so only the release branch is being deployed to the google cloud storage

MichelV96 committed:
Mysql url

MichelV96 committed:
Merge pull request #80 from xebia-research/feature/Support_logging_#28

Feature/support logging #28

nikro412956 committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/adding_GCS_from_Xebia_#64

Commit:
Added produces attribute to api calls again to fix e2e test

nikro412956 committed:
checking the variable values from travis

MichelV96 committed:
Merge branch 'develop' into feature/Support_configurable_settings_#27

nikro412956 committed:
local directory selection test for GCS deploy on travis

nikro412956 committed:
changed when travis should deploy so i can actually test if it can select the directories

nikro412956 committed:
try to deploy one directory

jordy1111100 committed:
Merge pull request #81 from xebia-research/feature/Support_configurable_settings_#27

Feature/support configurable settings #27

Commit:
fixed typo in html report

nikro412956 committed:
adding GCS from Xebia #64
made it so it will push everything from the release branch because i can't seem to get it to deploy with more then 1 local directory

nikro412956 committed:
Merge branch 'develop' of https://github.com/xebia-research/headerbuddy into feature/adding_GCS_from_Xebia_#64

MichelV96 committed:
Merge branch 'develop' into feature/Certificate_Details_#73

MichelV96 committed:
Merge branch 'feature/Certificate_Details_#73' of https://github.com/xebia-research/headerbuddy into feature/Certificate_Details_#73

anthony0tran committed:
Merge pull request #83 from xebia-research/feature/adding_GCS_from_Xebia_#64

Feature/adding gcs from xebia #64

Commit:
Merge branch 'develop' into feature/Certificate_Details_#73

Commit:
Merge pull request #84 from xebia-research/feature/Certificate_Details_#73

Feature/certificate details #73

Commit:
Merge branch 'develop' into feature/HTML_Response_#21

jordy1111100 committed:
Updated the seed 'Do' values with better descriptions

jordy1111100 committed:
Corrected descriptions for some header values

anthony0tran committed:
Merge pull request #85 from xebia-research/feature/HTML_Response_#21

Feature/html response #21

jordy1111100 committed:
Merge branch 'develop' into feature/Implement_seed_method_#24

jordy1111100 committed:
Made changes to the seed based on the review

anthony0tran committed:
Merge pull request #86 from xebia-research/feature/Implement_seed_method_#24

Feature/implement seed method #24

MichelV96 committed:
Delete codeStyleConfig.xml

Commit:
fixed category name and colors in html report

commjoen commented on pom.xml:
```diff
@@ -0,0 +1,301 @@
+
+    <groupId>com.xebia</groupId>
+    <artifactId>headerbuddy</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
```
Please make sure that your pom versioning is in line with your git versioning. This can be automated.

commjoen commented on src/main/java/com/xebia/headerbuddy/annotations/validators/ValidEmailValidator.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public void initialize(ValidEmail constraintAnnotation) {
+        // Left blank
```
Did you leave it blank intentionaly?

commjoen commented on src/main/java/com/xebia/headerbuddy/models/ApiKey.java:
```diff
@@ -0,0 +1,13 @@
+package com.xebia.headerbuddy.models;
+
+public class ApiKey {
+    private String key;
```
please make private members final if they are not volatile.

commjoen commented on src/main/java/com/xebia/headerbuddy/models/CertificateDetails.java:
```diff
@@ -0,0 +1,54 @@
+    private String publicKeyAlgorithm;
+    private String publicKeyFormat;
+    private Date expireDate;
+    private String issuer;
```
please make private members final if they are not volatile

MichelV96 replied:
final variables can only be assigned in the declaration, instance initializer or constructor. We are using a different setter. Reasoning is the way we get the certificate.

commjoen commented on src/main/java/com/xebia/headerbuddy/models/CertificateDetails.java:
```diff
@@ -0,0 +1,54 @@
+        this.type = certificate.getType();
+        this.hashCode = certificate.hashCode();
+        this.publicKeyAlgorithm = certificate.getPublicKey().getAlgorithm();
+        this.publicKeyFormat = certificate.getPublicKey().getFormat();
```
why are you using the this keyword here?

commjoen reviewed (commented):
First few comments, more comments will follow from both nanne and me...

jordy1111100 committed:
Added profile parameter

jordy1111100 committed:
Modified protocolHandler and extended seed

Commit:
changed minor things in html report, added back targetUrl

jordy1111100 committed:
Merge pull request #89 from xebia-research/feature/Support_profile_parameter_#87

Feature/support profile parameter #87

Commit:
Merge branch 'develop' into feature/Minor_Winnovation_fixes

MichelV96 committed:
Merge pull request #90 from xebia-research/feature/Minor_Winnovation_fixes

Feature/minor winnovation fixes

nikro412956 committed:
added a suppression file for the dependency checker because the vulnerability does not apply to this project.

nikro412956 committed:
reset the properties so that is can use the docker container again

nikro412956 committed:
Delete the dependency check report

deleted the dependency check report because it is auto generated on build.
i will also skip the travis ci pipeline check because the only thing changed is that this file is being deleted.
[skip ci]

nbaars commented on src/main/java/com/xebia/headerbuddy/annotations/validators/ValidMethodValidator.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.List;
+import java.util.Set;
+
+public class ValidMethodValidator implements ConstraintValidator<ValidMethod, String> {
```
Method is too long, try breaking it up in several methods with descriptive names which probably remove the need for the comments.

nbaars commented on src/test/java/com/xebia/headerbuddy/e2e/HeaderBuddyApiKeyTest.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Test
+    public void HeaderBuddyCorrectApiKeyTest() {
+        String url = "http://localhost:"+port+"/headerbuddy/api?key="+ key +"&url=" + testedUrl;
```
You can use  `String url = "/headerbuddy/api?key="+ key +"&url=" + testedUrl; ` 
Spring Boot wil automatically take care of the host and port

nbaars commented on src/test/java/com/xebia/headerbuddy/e2e/HeaderBuddyApiKeyTest.java:
```diff
@@ -0,0 +1,62 @@
+    private EuserRepository userRepository;
+
+    // Get the random port spring is running on
+    @Value("${local.server.port}")
```
See comment below this is not necessary if you change the call with the `TestRestTemplate`

nbaars commented:
# General

Collection of general comments which are not easy to put along side the source code.

## README

Please extend the README.md file with a short description what the project is about and how you can build it through the command line. 

## RequestBehaviour and subclasses

Looks like the only need for the sub classes is to change the method to be used in the connection:

```
public List<Header> doRequest() throws Exception {
        URL target = new URL(this.url);

        HttpURLConnection connection = (HttpURLConnection) target.openConnection();
        switch (this.methodName) {
            case "PATCH":
                connection.setRequestProperty("X-HTTP-Method-Override", "PATCH");
                connection.setRequestMethod("POST");
                break;
            case "CONNECT":
```

The object structure can be simplified a little bit if you create enums for the http methods and ask the enum for the request method, for example:

```
public enum HttpRequestMethod {

    GET("GET"), CONNECT("POST"); //add others as well....

    private final String requestMethod;

    HttpRequestMethod(String method) {
        this.requestMethod = method;
    }

    public String requestMethod() {
        return requestMethod;
    }
}
```

Now you can change the `RequestBehavior` class as follows:

```
  public RequestBehaviour(final String url, final HttpRequestMethod methodName) {
        this.url = url;
        this.method = methodName;
    }

    public List<Header> doRequest() throws Exception {
        URL target = new URL(this.url);
        HttpURLConnection connection = (HttpURLConnection) target.openConnection();
        connection.setRequestMethod(this.method.requestMethod());
        Map<String, List<String>> headerFields = connection.getHeaderFields();
        List<Header> headers = HeaderSerializer.serialize(headerFields, this.url);

        return headers;
    }
```

This way you can eliminate all the sub classes (which only play the role of calling the super class, no behavior is overridden). The super class asks a specific enum for the request method, if you add an extra http method you do not need to change `RequestBehaviour` class anymore, only add a new enum for a different http method.

## Structure of test cases

In the test package the files are divided in different packages which do not reflect the structure of the package in the `src/main/java` package. I would recommend to keep the tests in the same package as the original source file. The division between unit test/integration test follows from the class under test. A Spring controller will probably not be tested as a single unit but inside a Spring container.

## Testing 

You can easily use the H2 database (in memory) for testing, this way you do not need to setup a MySQL database in order to run the tests.
To enable this add the following dependency to the `pom.xml`:

```
<dependency>
  <groupId>com.h2database</groupId>
  <artifactId>h2</artifactId>
  <scope>test</scope>
</dependency>
```

And add an `application.properties` to the `test/resources` directory:

```
spring.jpa.hibernate.ddl-auto=create

spring.datasource.driver-class-name=org.h2.Driver


#Program settings
key.required = true
webcrawler.limit = 20

#Server settings
server.port = 8080
mysql.port = 3306

#Test settings
test.http.url = http://andonoz.com
test.https.url = https://xebia.com
```
You can also override the driver-class and enable a profile in your testcases.
With this changes Hibernate will create an in-memory database while running the tests

nbaars commented on src/main/java/com/xebia/headerbuddy/HeaderbuddyApplication.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.context.annotation.ComponentScan;
+
+@SpringBootApplication
+@ComponentScan
```
`@ComponentScan` can be removed, implicit by using `SpringBootApplication`

nbaars commented on src/main/java/com/xebia/headerbuddy/SwaggerConfig.java:
```diff
@@ -0,0 +1,40 @@
+    public Docket api() {
+        return new Docket(DocumentationType.SWAGGER_2)
+                .select()
+                .apis(RequestHandlerSelectors.basePackage("com.xebia.headerbuddy.controllers"))
```
Easier to use:                 `.apis(RequestHandlerSelectors.basePackage(HeaderBuddyController.class.getPackage().getName()))`

more "refactor safe"

nikro412956 committed:
forcing a travis pipeline check

the travis pipeline won't run for the pullrequest and this should fix that because the skip i had in the previous commit description messed it up

MichelV96 committed:
Controller methods to post + response from get to post

MichelV96 committed:
Merge remote-tracking branch 'origin/develop' into feature/switching_POST_#91

MichelV96 committed:
mysql string

MichelV96 committed:
Added dependency report to gitignore

OldaranT committed:
Merge branch 'release/Winnovation-1.0.0'

OldaranT committed:
Merge tag 'Winnovation-1.0.0' into develop

nikro412956 committed:
creating this branch on github and testing if our application works in java 9

MichelV96 committed:
HTML bug fix now it returns html in a java package as well

OldaranT committed:
Merge pull request #95 from xebia-research/feature/switching_POST_#91

Feature/switching post #91

anthony0tran committed:
Wrote README.md

anthony0tran committed:
Merge branch 'develop' into feature/Make_the_Readme_#96

anthony0tran committed:
Merge branch 'develop' into feature/Resolve_vulnerable_dependencies_#75

anthony0tran committed:
Update .travis.yml

- Minor change, removed oraclejdk9

anthony0tran committed:
Update README.md

- Added wait time
- minor fixes

nikro412956 committed:
Merge pull request #93 from xebia-research/feature/Resolve_vulnerable_dependencies_#75

Feature/resolve vulnerable dependencies #75

anthony0tran committed:
Update README.md

- added some notes
- changed optional

anthony0tran committed:
Update README.md

- Forgot capital letters

anthony0tran committed:
Update README.md

- added another note

nikro412956 committed:
Merge branch 'develop' into feature/Make_the_Readme_#96

nikro412956 committed:
Merge pull request #97 from xebia-research/feature/Make_the_Readme_#96

Feature/make the readme #96

nikro412956 committed:
resolved a naming issue which caused the travis pipeline not to deploy to the GCS bucket.

OldaranT committed:
Temporly fix for html/css in windows.

MichelV96 committed:
Merge pull request #98 from xebia-research/feature/adding_GCS_from_Xebia_#64

resolved a naming issue

MichelV96 committed:
Merge branch 'develop' into feature/switching_POST_#91

MichelV96 committed:
Merge pull request #99 from xebia-research/feature/switching_POST_#91

Temporly fix for html/css in windows.

MichelV96 committed:
Feature/code review changes #94 (#100)

* changes made to models, attributes

* Added in memory database to test cases

* Checkstyle error

* -

* RequestBehavior changed -> enums instead of classes; changed tests and validator aswell

* Travis debug

* Mysql password correction

* Test structure changed

* Docker file version

* Docker file version

JChavannes01 committed:
Edited the travis config to prevent auth plugin error (#102)
